### PR TITLE
Added Baidu Maps RPC Fixes #2158

### LINKED ISSUE #2158: Baidu Maps API
Had a request to use NetsBlox in China, but Google services will not work there. The first step to allowing it to be used in places where Google services are unavailble is to provide alternative mapping services, as many of the best NetsBlox examples rely on Google Maps right now.

### PR DESCRIPTION
**Pull request by gsteinLTU**
This is the existing Google Maps RPC (which admittedly may receive changes as part of #2155 that will also be applied to this) modified to use Baidu maps instead of Google Maps.

baidumaps

I can provide an API key for use in testing.

### CODE REVIEW

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+const cache = CacheManager.caching({store: 'memory', max: 1000, ttl: Infinity});
+const key = process.env.BAIDU_MAPS_KEY;
+
+var storage;
+
+// Retrieving a static map image
+var baseUrl = 'https://api.map.baidu.com/staticimage/v2?mcode=666666&',
+    getStorage = function() {
+        if (!storage) {
+            storage = Storage.create('baidu-staticmaps');
```
Comment (brollb) [inline src/server/rpc/procedures/baidu-maps/baidu-maps.js]: maybe `baidu-maps` might be better. I am guessing this is a result of copying from the google maps service (which used to be called "staticmaps").

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    // new latlon in px
+    let targetPx = merc.px([lon, lat], map.zoom);
+    // difference in px
+    return {x: (targetPx[0] - curPx[0]), y: -(targetPx[1] - curPx[1])};
+};
+
+
+// precisionLimit if present would limit the precision of coordinate parameters
+BaiduMaps.prototype._getBaiduParams = function(options, precisionLimit) {
+    // Create the params for Google
```
Comment (brollb) [inline src/server/rpc/procedures/baidu-maps/baidu-maps.js]: (not for Google)

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+        });
+
+    });
+};
+
+/**
+ * Get a map image of the given region.
+ * @param {Latitude} latitude Latitude of center point
+ * @param {Longitude} longitude Longitude of center point
+ * @param {BoundedNumber<0,1024>} width Image width
```
Comment (brollb) [inline src/server/rpc/procedures/baidu-maps/baidu-maps.js]: Minimum value should probably be 1

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+ * @param {BoundedNumber<0,1024>} height Image height
```
Comment (brollb) [inline src/server/rpc/procedures/baidu-maps/baidu-maps.js]: Minimum value should probably be 1

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+};
+
+/**
+ * Convert x value of map image to longitude.
+ * @param {Number} x x value of map image
+ * @returns {Longitude} Longitude of the x value from the image
+ *
+ * @deprecated
+ */
+BaiduMaps.prototype.getLongitude = function(x){
```
Comment (brollb) [inline src/server/rpc/procedures/baidu-maps/baidu-maps.js]: We don't need to support deprecated methods for BaiduMaps

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    return this.getLongitudeFromX(x);
+};
+
+/**
+ * Convert y value of map image to latitude.
+ * @param {Number} y y value of map image
+ * @returns {Latitude} Latitude of the y value from the image
+ *
+ * @deprecated
```
Comment (brollb) [inline src/server/rpc/procedures/baidu-maps/baidu-maps.js]: Same as above

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+BaiduMaps.isSupported = () => {
+    if(!key){
+        /* eslint-disable no-console*/
+        console.error('BAIDU_MAPS_KEY is missing.');
+        /* eslint-enable no-console*/
+    }
+    return !!key;
+};
+// Map of argument name to old field name
+BaiduMaps.COMPATIBILITY = {
```
Comment (brollb) [inline src/server/rpc/procedures/baidu-maps/baidu-maps.js]: None of the compatibility options are needed. In the GoogleMaps service, these are only for backwards-compatibility.

File: test/unit/server/rpc/procedures/baidu-maps.spec.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
+describe('baidu-staticmaps', function() {
```
Comment (brollb) [inline test/unit/server/rpc/procedures/baidu-maps.spec.js]: `baidu-maps` would be better.

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js (hunk 1 above)
Comment (gsteinLTU) [reply src/server/rpc/procedures/baidu-maps/baidu-maps.js]: Staticmaps is their official term for it, is it worth changing?

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js (hunk 2 above)
Comment (gsteinLTU) [reply src/server/rpc/procedures/baidu-maps/baidu-maps.js]: Fixed now

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js (hunk 3 above)
Comment (gsteinLTU) [reply src/server/rpc/procedures/baidu-maps/baidu-maps.js]: Min values are fixed now.

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js (hunk 5 above)
Comment (gsteinLTU) [reply src/server/rpc/procedures/baidu-maps/baidu-maps.js]: They're gone now.

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js (hunk 7 above)
Comment (gsteinLTU) [reply src/server/rpc/procedures/baidu-maps/baidu-maps.js]: Are they used in the existing Google Maps examples?

Comment (brollb) [reply src/server/rpc/procedures/baidu-maps/baidu-maps.js]: I am not sure but they are likely used within other user projects in the database.

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+
+const logger = require('../utils/logger')('baidu-staticmaps');
+const request = require('request');
+const SphericalMercator = require('sphericalmercator');
+const geolib = require('geolib');
+const merc = new SphericalMercator({size:256});
+const CacheManager = require('cache-manager');
+const Storage = require('../../storage');
+
+// TODO: Change this cache to mongo or something (file?)
```
Comment (hamidzr) [inline src/server/rpc/procedures/baidu-maps/baidu-maps.js]: this library supports other stores like filesystem store `cache-manager-fs` as is used in `ApiConsumer` so switching should be simple if a persistent storage is beneficial.

File: test/unit/server/rpc/procedures/baidu-maps.spec.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
```
Comment (hamidzr) [inline test/unit/server/rpc/procedures/baidu-maps.spec.js]: if we want the two services to have the same API maybe we can have them share the tests.

**Comment by hamidzr** [comment]
I was wondering if it would be better to introduce Baidu as an additional "engine" to a uniform map service which would talk to different map providers (Google, Baidu, OpenStreet..) or not.

File: src/server/rpc/procedures/baidu-maps/baidu-maps.js (hunk 1 above)
Comment (gsteinLTU) [reply src/server/rpc/procedures/baidu-maps/baidu-maps.js]: It's changed now, good point that the service name doesn't match.

**Comment by gsteinLTU** [comment]
@hamidzr Maybe that change could be rolled into #2155 with the other changes to the Google Maps RPC. The Baidu, MapQuest OpenStreetMap (https://developer.mapquest.com/documentation/open/static-map-api/v4/) and Google APIs are really similar.

**Comment by brollb** [comment]
I agree that the amount of code duplication suggests that there might be a better structure for the code (such as a base class for map providers). Of course, these should only be sharing logic internally and should still be distinct services for the user IMO.

However, since #2155 will already be pretty big, I wouldn't roll anything extra into it. Making it much bigger will make the review more time-consuming. That being said, adding the engine/base class/whatever in another PR sounds like a good idea to me to keep the size of the individual PRs minimal and making the review easier (and hopefully less risk of introducing bugs).

**Comment by brollb** [comment]
Could you confirm that there aren't any weird things with the coordinate system? It looks like China uses a different coordinate system (BD-09): https://en.wikipedia.org/wiki/Baidu_Maps

Other than that, LGTM

**Comment by brollb** [comment]
@gsteinLTU Have you had a chance to confirm that the BD-09 coordinate system will not cause problems with this service's reuse of the conversion utilities from the GoogleMaps service?

**Pull request closed without merging**